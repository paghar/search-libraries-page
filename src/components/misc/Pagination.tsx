import {useMemo} from "react";
import RightArrow from "./svg/RightArrow";
import LeftArrow from "./svg/LeftArrow";

const DOTS = "...";

type Props = {
  currentPage: number;
  totalCount: number;
  perPage: number;
  onClick: (currenPage: number) => void;
};
const Pagination: React.FC<Props> = ({currentPage, totalCount, perPage, onClick}) => {
  const totalPageCount = Math.ceil(totalCount / perPage);

  const range = (start: number, end: number) => {
    const length = end - start + 1;
    return Array.from({length}, (_, idx) => idx + start);
  };

  const paginationRange = useMemo(() => {
    const siblingCount = 1;

    const totalPageNumbers = siblingCount + 5;

    if (totalPageNumbers >= totalPageCount) {
      return range(1, totalPageCount);
    }

    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPageCount);

    const shouldShowLeftDots = leftSiblingIndex > 2;
    const shouldShowRightDots = rightSiblingIndex < totalPageCount - 2;

    const firstPageIndex = 1;
    const lastPageIndex = totalPageCount;

    if (!shouldShowLeftDots && shouldShowRightDots) {
      const leftItemCount = 3 + 2 * siblingCount;
      const leftRange = range(1, leftItemCount);

      return [...leftRange, DOTS, totalPageCount];
    }

    if (shouldShowLeftDots && !shouldShowRightDots) {
      const rightItemCount = 3 + 2 * siblingCount;
      const rightRange = range(totalPageCount - rightItemCount + 1, totalPageCount);
      return [firstPageIndex, DOTS, ...rightRange];
    }

    if (shouldShowLeftDots && shouldShowRightDots) {
      const middleRange = range(leftSiblingIndex, rightSiblingIndex);
      return [firstPageIndex, DOTS, ...middleRange, DOTS, lastPageIndex];
    }
  }, [totalCount, perPage, currentPage]);

  const onNext = () => {
    if (currentPage === totalPageCount) return;
    onClick(currentPage + 1);
  };

  const onPrevious = () => {
    if (currentPage === 1) return;
    onClick(currentPage - 1);
  };

  if (currentPage === 0 || totalPageCount < 2 || !paginationRange) {
    return null;
  }
  return (
    <ul className="flex justify-center items-center gap-2">
      {/* Left navigation arrow */}
      <li 
        role="prevButton"     
        className={`p-3 border rounded-md hover:bg-black-600 group transition-all cursor-pointer ${
          currentPage === 1 && "border-pink-400 [&>svg]:fill-pink-400"
        }`}
        onClick={onPrevious}
      >
        <LeftArrow className="group-hover:fill-gray-500"/>
       
      </li>
      {paginationRange?.map((pageNumber,index) => {
        // If the pageItem is a DOT, render the DOTS unicode character
        if (pageNumber === DOTS) {
          return (
            <li key={`item-number${index}`} className=" dots">
              &#8230;
            </li>
          );
        }

        // Render our Page Pills
        return (
          <li            
            key={pageNumber}
            className={`text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer ${
              pageNumber === currentPage && "text-gray-500 bg-pink-600"
            }`}
            onClick={() => onClick(+pageNumber)}
          >
            {pageNumber}
          </li>
        );
      })}
      {/*  Right Navigation arrow */}
      <li    
        role="nextButton"      
        className={`p-3  border rounded-md hover:bg-black-600 group transition-all cursor-pointer ${
          currentPage === totalPageCount && "border-pink-400 [&>svg]:fill-pink-400"
        }`}
        onClick={onNext}
      >
        <RightArrow  className=" group-hover:fill-white-500"/>      
      </li>
    </ul>
  );
};

export default Pagination;
