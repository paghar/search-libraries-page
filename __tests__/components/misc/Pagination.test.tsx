import React from "react";
import {render, fireEvent} from "@testing-library/react";
import Pagination from "../../../src/components/misc/Pagination";

const currentPage = 3;
const totalCount = 50;
const perPage = 10;
const onClick = jest.fn();

describe("Pagination Component", () => {

  const {container} = render(
    <Pagination currentPage={currentPage} totalCount={totalCount} perPage={perPage} onClick={onClick} />
  );

  it("Renders Snapshot Buttons", () => {    
    expect(container).toMatchSnapshot();
  });

});

describe("Pagination Component", () => {
  test("renders correct page numbers and handles navigation clicks", () => {   

    const {getByText, getByRole} = render(
      <Pagination currentPage={currentPage} totalCount={totalCount} perPage={perPage} onClick={onClick} />
    );

    // Check if page numbers and arrows are rendered
    const pageNumbers = [1,2, 3, 4, 5];
    pageNumbers.forEach(page => {      
      expect(getByText(page.toString())).toBeInTheDocument();   
    });

    // Find buttons for navigation
    const prevButton = getByRole("prevButton");
    // Simulate a click on the previous arrow
    fireEvent.click(prevButton);
    expect(onClick).toHaveBeenCalledWith(currentPage - 1);

    // Find buttons for navigation
    const nextButton = getByRole("nextButton");
    // Simulate a click on the next arrow
    fireEvent.click(nextButton);
    expect(onClick).toHaveBeenCalledWith(currentPage + 1);

  });
});