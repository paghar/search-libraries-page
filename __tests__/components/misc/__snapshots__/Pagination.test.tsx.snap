// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component Renders Snapshot Buttons 1`] = `
<div>
  <ul
    class="flex justify-center items-center gap-2"
  >
    <li
      class="p-3 border rounded-md hover:bg-black-600 group transition-all cursor-pointer false"
      role="prevButton"
    >
      <svg
        class="group-hover:fill-gray-500"
        height="20"
        viewBox="0 0 320 512"
        width="20"
      >
        <path
          d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
        />
      </svg>
    </li>
    <li
      class="text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer false"
    >
      1
    </li>
    <li
      class="text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer false"
    >
      2
    </li>
    <li
      class="text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer text-gray-500 bg-pink-600"
    >
      3
    </li>
    <li
      class="text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer false"
    >
      4
    </li>
    <li
      class="text-lg px-4 py-2 border rounded-md hover:bg-black-600 hover:text-gray-500 transition-all cursor-pointer false"
    >
      5
    </li>
    <li
      class="p-3  border rounded-md hover:bg-black-600 group transition-all cursor-pointer false"
      role="nextButton"
    >
      <svg
        class=" group-hover:fill-white-500"
        height="20"
        viewBox="0 0 320 512"
        width="20"
      >
        <path
          d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
        />
      </svg>
    </li>
  </ul>
</div>
`;
